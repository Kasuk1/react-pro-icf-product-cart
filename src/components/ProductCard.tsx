import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  IProductContext,
  IProduct,
  IOnChangeArgs,
  IInitialvalues,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';
import { IProductCardHandlers } from '../interfaces/interfaces';

export const ProductContext = createContext({} as IProductContext);
const { Provider } = ProductContext;
export interface IProductCard {
  //children?: ReactElement | ReactElement[];
  children: (args: IProductCardHandlers) => JSX.Element;
  product: IProduct;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: IOnChangeArgs) => void;
  value?: number;
  initialValues?: IInitialvalues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: IProductCard) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    product,
    onChange,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          maxCount: initialValues?.maxCount,
          isMaxCountReached,
          reset,
          increaseBy,
          product,
        })}
      </div>
    </Provider>
  );
};
