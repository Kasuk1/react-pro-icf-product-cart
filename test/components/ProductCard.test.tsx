import * as React from 'react';
import renderer from 'react-test-renderer';
import ProductCard from '../../src/components';
import { product1 } from '../data/products';

const { act } = renderer;

describe('<ProductCard /> Tests', () => {
  test('should render correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {({}) => <h1>Product children</h1>}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should increment the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {({ count, increaseBy }) => (
          <>
            <h1>Product Card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}>Increment</button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
