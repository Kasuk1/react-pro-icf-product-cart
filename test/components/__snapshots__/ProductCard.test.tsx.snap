// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> Tests should increment the counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Increment
  </button>
</div>
`;

exports[`<ProductCard /> Tests should render correctly 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product children
  </h1>
</div>
`;
